### ISSUE TITLE
Config Serialization creates "too exact" config

### ISSUE THREAD

**Issue by Atticus1806** [rwth-i6/returnn_common#129]
First of all, I really like the new updates pushed yesterday to generate a config. Its good to understand and makes config generation now a lot more straight forward! 
Nevertheless I started playing around with it in an example taken from the README (with modfications but I will PR them in a different PR), and realized layers like `LayerNorm` generate quite extensive config code. 

My example is:
```
class MyModelBlock(nn.Module):
  def __init__(self, dim: nn.Dim, hidden: nn.Dim, dropout: float = 0.1):
    super().__init__()
    self.dropout_dim = dim
    self.layer_norm = nn.LayerNorm()
    self.linear_out = nn.Linear(dim)
    self.linear_hidden = nn.Linear(hidden)
    self.dropout = dropout

  def __call__(self, x: nn.Tensor) -> nn.Tensor:
    y = self.layer_norm(x)
    y = self.linear_hidden(y)
    y = nn.sigmoid(y)
    y = self.linear_out(y)
    y = nn.dropout(y, dropout=self.dropout, axis=[nn.batch_dim])

    return x + y

def get_config():
  nn.reset_default_root_name_ctx()
  input_dim = nn.FeatureDim("input", 10)
  target_dim = nn.FeatureDim("target", 7)
  data = nn.get_extern_data(nn.Data("data", dim_tags=[nn.batch_dim, input_dim]))
  block = MyModelBlock(input_dim, target_dim)
  out = block(data)
  out.mark_as_default_output()
  config_code = nn.get_returnn_config().get_complete_py_code_str(block)
  print(config_code)
```

which then prints (exemplary for `layer_norm`):

```
  'layer_norm': {
    'class': 'subnetwork',
    'from': [],
    'subnetwork': {
      'reduce': {
        'class': 'reduce',
        'from': 'base:data:data',
        'mode': 'mean',
        'axis': input_dim,
        'out_shape': {batch_dim, time_dim}
      },
      'squared_difference': {
        'class': 'combine',
        'from': ['base:data:data', 'reduce'],
        'kind': 'squared_difference',
        'out_shape': {batch_dim, time_dim, input_dim}
      },
      'variance': {
        'class': 'reduce',
        'from': 'squared_difference',
        'mode': 'mean',
        'axis': input_dim,
        'out_shape': {batch_dim, time_dim}
      },
      'sub': {
        'class': 'combine',
        'from': ['base:data:data', 'reduce'],
        'kind': 'sub',
        'out_shape': {batch_dim, time_dim, input_dim}
      },
      'constant': {'class': 'constant', 'value': 1e-06},
      'add': {
        'class': 'combine',
        'from': ['variance', 'constant'],
        'kind': 'add',
        'out_shape': {batch_dim, time_dim}
      },
      'rsqrt': {
        'class': 'activation',
        'from': 'add',
        'activation': 'rsqrt',
        'out_shape': {batch_dim, time_dim}
      },
      'mul': {
        'class': 'combine',
        'from': ['sub', 'rsqrt'],
        'kind': 'mul',
        'out_shape': {batch_dim, time_dim, input_dim}
      },
      'mul_0': {
        'class': 'combine',
        'from': ['mul', 'scale'],
        'kind': 'mul',
        'out_shape': {batch_dim, time_dim, input_dim}
      },
      'add_0': {
        'class': 'combine',
        'from': ['mul_0', 'bias'],
        'kind': 'add',
        'out_shape': {batch_dim, time_dim, input_dim}
      },
      'output': {
        'class': 'copy',
        'from': 'add_0',
        'out_shape': {batch_dim, time_dim, input_dim}
      },
      'scale': {
        'class': 'variable',
        'shape': [
          input_dim
        ],
        'param_name': 'param',
        'init': 1.0
      },
      'bias': {
        'class': 'variable',
        'shape': [
          input_dim
        ],
        'param_name': 'param',
        'init': 0.0
      }
    },
    'out_shape': {batch_dim, time_dim, input_dim}
  },
```

Is this intended like this?

**Comment by albertz** [comment]
It looks pretty straightforward, or not? What would you have expected? How should it be shorter than that?

**Comment by Atticus1806** [comment]
I feel like there are a lot of sublayers that make the whole network bloated. 
e.g. bias, scale in this case or for a linear layer:
```
  'linear_out': {
    'class': 'subnetwork',
    'from': [],
    'subnetwork': {
      'random': {
        'class': 'random',
        'shape': (
          target_dim,
          input_dim
        ),
        'distribution': 'uniform',
        'minval': -0.5940885257860046,
        'maxval': 0.5940885257860046,
        'dtype': 'float32',
        'static': True
      },
      'dot': {
        'class': 'dot',
        'from': ['base:sigmoid', 'weight'],
        'reduce': target_dim,
        'out_shape': {batch_dim, time_dim, input_dim}
      },
      'add': {
        'class': 'combine',
        'from': ['dot', 'bias'],
        'kind': 'add',
        'out_shape': {batch_dim, time_dim, input_dim}
      },
      'output': {
        'class': 'copy',
        'from': 'add',
        'out_shape': {batch_dim, time_dim, input_dim}
      },
      'weight': {
        'class': 'variable',
        'shape': [
          target_dim,
          input_dim
        ],
        'param_name': 'param',
        'init_by_layer': 'random'
      },
      'bias': {
        'class': 'variable',
        'shape': [
          input_dim
        ],
        'param_name': 'param',
        'init': 0.0
      }
    },
    'out_shape': {batch_dim, time_dim, input_dim}
  },
```
ofc. this makes everthing explicit but I am not sure if we need this level of detail for some "basic" layers (explicit bias, weight, random etc.). Or is there a way to generate a config with less detail and more implicit code?

**Comment by Atticus1806** [comment]
>What would you have expected? How should it be shorter than that?

Is there a reason not to use the `RETURNN` `Linear` or `LayerNorm` layer here?

**Comment by albertz** [comment]
> I feel like there are a lot of sublayers that make the whole network bloated.
> Is there a reason not to use the `RETURNN` `Linear` or `LayerNorm` layer here?

Well, this was discussed previously here and we came to the conclusion that we want to have all variables explicit. This makes many things much more straightforward (e.g. implementing weight norm or weight dropout in a generic way). This also was a very fundamental decision.

Also, we don't want to have the logic too complicated. So, while in theory we could maybe check if some subnetwork dict matches exactly the behavior like it would be if you would use a `LinearLayer`, such check would (1) be complicated and (2) make the code more complex and (3) easy to get wrong. We want to have the code simple and straightforward.

> Or is there a way to generate a config with less detail and more implicit code?

No. I'm not sure how you would even realize that. This would imply to have lots of duplicated code for "implicit shorter dicts" and the straightforward case, and many things would not work correctly for the "implicit code" case.

But as said, this was already discussed previously. As this was a very fundamental decision, it's difficult to change that now.

Having said all that, there are still some parts of the generated dict which we might optimize (cleanup/shorten) further:

* The random init of vars could maybe be separate? Or for the simple cases reduce to just an option? Not sure.
* We could avoid the `constant` layer maybe by using `eval` in the combine layer, or extending the `CombineLayer`, or introduce a new `CombineWithConstantLayer` or so.
* Maybe other things? Maybe "fuse" some arithmetics into an eval layer?

If you want to discuss any of these points, maybe open separate issues on each individual point because I think most of them are independent of each other.

However, as said before, I think we should not overcomplicate it. The code should stay relatively simple and straightforward.

Also, I think this is not such a high priority for now to really invest much time and energy on this.

**Comment by Atticus1806** [comment]
Thanks for the explanation though!
